apps/vaporgui/VaporWidgets.cpp

```diff
@@ -0,0 +1,206 @@
+#include <QFileDialog>
+
+#include "VaporWidgets.h"
+#include "FileOperationChecker.h"
+#include "ErrorReporter.h"
+
+#include "ui_VPushButton.h"
+#include "ui_VCheckBox.h"
+#include "ui_VComboBox.h"
+#include "ui_VPathSelector.h"
+
+#include <iostream>
+#include <cassert>
+
+VPushButton::VPushButton(
+        QWidget *parent,
+        std::string labelText,
+        std::string buttonText
+    ) :
+    QWidget(parent),
+    Ui_VPushButton()
+{
+    setupUi(this);
+
+    SetLabelText( QString::fromStdString( labelText ) );
+    SetButtonText( QString::fromStdString( buttonText ) );
+
+    connect( _myButton, SIGNAL( pressed() ),
+        this, SLOT( _buttonPressed() ) );
+}
+
+void VPushButton::SetLabelText( const std::string text )
+{
+    SetLabelText(QString::fromStdString( text ) );
+}
+
+void VPushButton::SetLabelText( const QString text )
+{
+    _myLabel->setText( text );
+}
+
+void VPushButton::SetButtonText( const std::string text )
+{
+    SetButtonText(QString::fromStdString( text ) );
+}
+
+void VPushButton::SetButtonText( const QString text )
+{
+    _myButton->setText( text );
+}
+
+void VPushButton::_buttonPressed() {
+    emit _pressed();
+}
+
+VComboBox::VComboBox(
+        QWidget *parent,
+        std::string labelText
+    ) :
+    QWidget(parent),
+    Ui_VComboBox()
+{
+    setupUi(this);
+
+    SetLabelText( labelText );
+
+    connect( _myCombo, SIGNAL( currentIndexChanged(int) ),
+        this, SLOT( _userIndexChanged(int) ) );
+}
+
+void VComboBox::SetLabelText( std::string text ) {
+    _myLabel->setText( QString::fromStdString(text) );
+}
+
+void VComboBox::_userIndexChanged(int index) {
+    emit _indexChanged(index);
+}
+
+int VComboBox::GetCurrentIndex() const {
+    return _myCombo->currentIndex();
+}
+
+std::string VComboBox::GetCurrentText() const {
+    return _myCombo->currentText().toStdString();
+}
+
+void VComboBox::AddOption( std::string option, int index) {
+    _myCombo->insertItem( index, QString::fromStdString(option) );
```

Comment:
what if `index` isn't a valid value?

Reply:
If index is negative, the item gets prepended.  If it's greater than the count, it gets appended.